types/time.go

```diff
@@ -207,159 +184,191 @@ func FromGoTime(t gotime.Time) MysqlTime {
+	t := ZeroCoreTime
+	t.setYear(uint16(year))
+	t.setMonth(uint8(month))
+	t.setDay(uint8(day))
+	t.setHour(uint8(hour))
+	t.setMinute(uint8(minute))
+	t.setSecond(uint8(second))
+	t.setMicrosecond(uint32(microsecond))
+	return t
 }
 
-// MaxMySQLTime returns Duration with maximum mysql time.
-func MaxMySQLTime(fsp int8) Time {
-	return NewTime(FromDate(0, 0, 0, TimeMaxHour, TimeMaxMinute, TimeMaxSecond, 0), mysql.TypeDuration, fsp)
+// FromDateChecked makes a internal time representation from the given date with field overflow check.
+func FromDateChecked(year, month, day, hour, minute, second, microsecond int) (CoreTime, bool) {
+	if uint64(year) >= (1 << (yearBitFieldStart - yearBitFieldEnd + 1)) {
+		return ZeroCoreTime, false
+	}
+	if uint64(month) >= (1 << (monthBitFieldStart - monthBitFieldEnd + 1)) {
+		return ZeroCoreTime, false
+	}
+	if uint64(day) >= (1 << (dayBitFieldStart - dayBitFieldEnd + 1)) {
+		return ZeroCoreTime, false
+	}
+	if uint64(hour) >= (1 << (hourBitFieldStart - hourBitFieldEnd + 1)) {
+		return ZeroCoreTime, false
+	}
+	if uint64(minute) >= (1 << (minuteBitFieldStart - minuteBitFieldEnd + 1)) {
+		return ZeroCoreTime, false
+	}
+	if uint64(second) >= (1 << (secondBitFieldStart - secondBitFieldEnd + 1)) {
+		return ZeroCoreTime, false
+	}
+	if uint64(microsecond) >= (1 << (microsecondBitFieldStart - microsecondBitFieldEnd + 1)) {
+		return ZeroCoreTime, false
+	}
+	return FromDate(year, month, day, hour, minute, second, microsecond), true
 }
 
-// Clock returns the hour, minute, and second within the day specified by t.
-func (t Time) Clock() (hour int, minute int, second int) {
-	return t.time.Hour(), t.time.Minute(), t.time.Second()
-}
+// coreTime is an alias to CoreTime, embedd in Time.
+type coreTime = CoreTime
 
 // Time is the struct for handling datetime, timestamp and date.
-// TODO: check if need a NewTime function to set Fsp default value?
 type Time struct {
-	time MysqlTime
-	tp   uint8
-	// Fsp is short for Fractional Seconds Precision.
-	// See http://dev.mysql.com/doc/refman/5.7/en/fractional-seconds.html
-	fsp int8
+	coreTime
 }
 
-// NewTime constructs time from core time, type and fsp.
-func NewTime(mt MysqlTime, tp uint8, fsp int8) Time {
-	return Time{
-		time: mt,
-		tp:   tp,
-		fsp:  fsp,
-	}
-}
-
-// Year returns year value.
-func (t Time) Year() int {
-	return t.time.Year()
-}
-
-// Month returns month value.
-func (t Time) Month() int {
-	return t.time.Month()
-}
-
-// Day returns day value.
-func (t Time) Day() int {
-	return t.time.Day()
+// Clock returns the hour, minute, and second within the day specified by t.
+func (t Time) Clock() (hour int, minute int, second int) {
+	return t.Hour(), t.Minute(), t.Second()
 }
 
-// Hour returns hour value.
-func (t Time) Hour() int {
-	return t.time.Hour()
-}
+const (
+	// Core time bit fields.
+	yearBitFieldStart, yearBitFieldEnd               uint64 = 63, 50
```

Comment:
IMO, the `yearBitFieldEnd` less than `yearBitFieldStart` look like weired.

Reply:
Seems the `yearBitFieldStart` just be used in the function `FromDateChecked` (`uint64(year) >= (1 << (yearBitFieldStart - yearBitFieldEnd + 1))`). Prefer to define `yearBitFieldOffset = 50` and `yearBitFieldWidth = 13`. The same changes can be applied to other constants.